packages/components/src/autocomplete/index.js

```diff
@@ -0,0 +1,336 @@
+import { withInstanceId, compose } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import List from './list';
+import Tags from './tags';
+import SearchControl from './control';
+
+/**
+ * A search box which filters options while typing,
+ * allowing a user to select from an option from a filtered list.
+ */
+export class Autocomplete extends Component {
+	static getInitialState() {
+		return {
+			filteredOptions: [],
+			selectedIndex: 0,
+			query: '',
+		};
+	}
+
+	constructor( props ) {
+		super( props );
+		this.state = this.constructor.getInitialState();
+
+		this.bindNode = this.bindNode.bind( this );
+		this.search = this.search.bind( this );
+		this.selectOption = this.selectOption.bind( this );
+		this.updateSelectedIndex = this.updateSelectedIndex.bind( this );
+	}
+
+	bindNode( node ) {
+		this.node = node;
+	}
+
+	reset( selected = this.props.selected ) {
+		const { multiple } = this.props;
+		const initialState = this.constructor.getInitialState();
+
+		// Reset to the option label if not using tags.
+		if ( ! multiple && selected.length && selected[ 0 ].label ) {
+			initialState.query = selected[ 0 ].label;
+		}
+
+		this.setState( initialState );
+	}
+
+	handleFocusOutside() {
+		this.reset();
+	}
+
+	isExpanded() {
+		const { filteredOptions, query } = this.state;
+
+		return filteredOptions.length > 0 || query;
+	}
+
+	hasTags() {
+		const { multiple, selected } = this.props;
+
+		if ( ! multiple ) {
+			return false;
+		}
+
+		return selected.some( item => Boolean( item.label ) );
+	}
+
+	selectOption( option ) {
+		const { multiple, onChange, selected } = this.props;
+		const { query } = this.state;
+		const newSelected = multiple ? [ ...selected, option ] : [ option ];
+
+		// Check if this is already selected
+		const isSelected = findIndex( selected, { key: option.key } );
+		if ( -1 === isSelected ) {
+			onChange( newSelected, query );
+		}
+
+		this.reset( newSelected );
+	}
+
+	updateSelectedIndex( value ) {
+		this.setState( { selectedIndex: value } );
+	}
+
+	announce( filteredOptions ) {
+		const { debouncedSpeak } = this.props;
+		if ( ! debouncedSpeak ) {
+			return;
+		}
+		if ( !! filteredOptions.length ) {
+			debouncedSpeak(
+				sprintf(
+					_n(
+						'%d result found, use up and down arrow keys to navigate.',
+						'%d results found, use up and down arrow keys to navigate.',
+						filteredOptions.length,
+						'woocommerce-admin'
+					),
+					filteredOptions.length
+				),
+				'assertive'
+			);
+		} else {
+			debouncedSpeak( __( 'No results.', 'woocommerce-admin' ), 'assertive' );
+		}
+	}
+
+	getFilteredOptions( query ) {
+		const { excludeSelectedOptions, getSearchExpression, maxResults, onFilter, options, selected } = this.props;
+		const selectedKeys = selected.map( option => option.key );
+		const filtered = [];
+
+		// Create a regular expression to filter the options.
+		const expression = getSearchExpression( escapeRegExp( query.trim() ) );
```

Comment:
On inline tags with the options always visible, pressing the escape key results in an error `index.js:126 Uncaught TypeError: Cannot read property 'trim' of null`.

Tab should also allow me to move on, but using only the keyboard you can get stuck. See the other comment.